src/cpufreq.c

```diff
@@ -72,66 +260,241 @@ static void cpufreq_submit (int cpu_num, double value)
-						sizeof (errbuf)));
-			fclose (fp);
-			return (-1);
+			if ((fp = fopen (filename, "r")) == NULL)
+			{
+				// cpu could be just switched off
+				// let's check the next one
+				continue;
+			}
+
+			if (fgets (buffer, sizeof(buffer), fp) == NULL)
+			{
+				char errbuf[1024];
+				WARNING ("cpufreq: fgets: %s",
+					 sstrerror (errno, errbuf,
+						    sizeof (errbuf)));
+				fclose (fp);
+				return (-1);
+			}
+
+			if (fclose (fp))
+			{
+				char errbuf[1024];
+				WARNING ("cpufreq: fclose: %s",
+					 sstrerror (errno, errbuf,
+						    sizeof (errbuf)));
+			}
+
+			/* You're seeing correctly: The file is reporting kHz values.. */
+			val = atoll (buffer) * 1000;
+
+			cpufreq_submit_current_value (i, val);
 		}
+	} /* end of if ( !report_distribution ) */
+
+	else // stats on CPU frequency distribution is used
+	{
+		char *fields[2];
+		char statind[64];
+		int numfields;
+		long int hertz;
+		derive_t time;
+		size_t last_index = hertz_size - 1;
+
+		if (!report_by_cpu)
+			for (size_t j=0; j < hertz_size; ++j)
+				time_all_cpus[j] = 0;
 
-		if (fclose (fp))
+		for (size_t i=0; i < num_cpu; ++i)
 		{
-			char errbuf[1024];
-			WARNING ("cpufreq: fclose: %s",
-					sstrerror (errno, errbuf,
-						sizeof (errbuf)));
-		}
+			status = ssnprintf (filename, sizeof (filename),
+					    "/sys/devices/system/cpu/cpu%zu/cpufreq/stats/time_in_state",
+					    i);
+
+			if ((status < 1) || ((size_t)status >= sizeof (filename)))
+			{
+				WARNING("cpufreq plugin: error in ssnprintf");
+				return (-1);
+			}
+
+			if ((fp = fopen (filename, "r")) == NULL)
+			{
+				// CPU could be just switched off
+				// let's check the next one
+				continue;
+			}
+
+			while (fgets(buffer, sizeof(buffer), fp) != NULL)
+			{
+				numfields = strsplit (buffer, fields, STATIC_ARRAY_SIZE(fields));
+
+				if (numfields != 2) // supported format contains 2 fields
+				{
+					// if its an empty line, just ignore line
+					if ( numfields == 0 ) continue;
+
+					WARNING ("cpufreq: wrong number of items on a line in %s", filename);
+					fclose (fp);
+					return (-1);
+				}
 
+				hertz = atol (fields[0]);
 
-		/* You're seeing correctly: The file is reporting kHz values.. */
-		val = atoll (buffer) * 1000;
+				time = atoll (fields[1]);
+				time *= 10; // cpufreq-stat module reports in 10ms
 
-		cpufreq_submit (i, val);
+				if (hertz <= 0)
+				{
+					WARNING ("cpufreq: something is wrong in %s: hertz = %ld", filename, hertz);
+					WARNING ("cpufreq: line in question %s %s", fields[0], fields[1]);
+					fclose (fp);
+					return (-1);
+				}
+
+				if (report_by_cpu)
+				{
+					status = ssnprintf (statind, sizeof (statind),
+										"%zu", i);
+
+					if ((status < 1) || ((size_t)status >= sizeof (statind)))
+					{
+						WARNING("cpufreq plugin: error in ssnprintf");
+						fclose (fp);
+						return (-1);
+					}
+
+					cpufreq_submit_distribution_value (statind, hertz, time);
+				}
+				else
+				{
+					int found = 0;
```

Comment:
- Make this a _Bool?
- Move it into the else clause since it's only used there.